crates/sui-types/src/messages.rs

```diff
@@ -96,6 +96,29 @@ pub struct TransferSui {
     pub amount: Option<u64>,
 }
 
+/// Gas will be paid from input coins, the remaining coins will
+/// all be paid to recipient.
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+pub struct PayAllSui {
+    /// The coins to be used for payment.
+    pub coins: Vec<ObjectRef>,
+    /// The address that will receive payment
+    pub recipient: SuiAddress,
+}
+
+/// Pay each recipient the corresponding amount using the input coins
+/// Gas will be paid by the remaining coins under sender.
```

Comment:
Please explain in detail the role of the gas object in this transaction (similar in PayAllSui). For instance, is the first object in the inputs always the same as the gas object? And is the first object used both for gas payment, AND also paying some amount to a recipient?

Reply:
> is the first object in the inputs always the same as the gas object? And is the first object used both for gas payment, AND also paying some amount to a recipient?

yes the first object in the inputs always the same as the gas object. 
the original SUI in the first coin can be used for both gas payment and for payment to recipients; in fact as they are all fungible SUIs, what matters here are: 1) the object ID of the first object and 2) the total residual SUI out of all input Sui tokens after paying total amount to recipients, as the total  residual SUI will be accumulated into the first input object.

will add a more detailed desc. here.